### ISSUE TITLE
Native CI for osx-arm64 (run tests, native compilation, ...)

### ISSUE THREAD

**Issue by h-vetinari** [conda-forge/conda-forge.github.io#1781]
Things work pretty well for M1's with all the work that's gone into cross-compilation (and the daily(?) job testing published artefacts), but it would still be nice to be able to test the packaged artifacts on the feedstocks before publishing them.

Adding M1 agents is now on the github roadmap (https://github.com/github/roadmap/issues/528), so perhaps once deployed this could be an interesting way to test the packages. conda-forge already uses GHA's for various things, but not to my knowledge on any feedstock, so this would probably need some work on smithy.

The roadmap issue isn't yet assigned to a specific quarter in the GH roadmap, so it might still be a while, unfortunately. But then, maybe other CI providers like azure end up being faster with offering osx-arm. In any case, I thought I'd open this issue for tracking.

**Comment by BastianZim** [comment]
> The roadmap issue isn't yet assigned to a specific quarter in the GH roadmap,

I reached out to GitHub support when they added it to the roadmap, and they plan it by the end of the financial year 2022.

**Comment by jaimergp** [comment]
The beta is now scheduled for Q3 2023:

<img width="444" alt="image" src="https://user-images.githubusercontent.com/2559438/207836841-0f0d0884-1ad9-4519-9548-3996b3733f86.png">

I found these providers already (or soonish) available, in case someone wants to experiment:

* https://cirrus-ci.org/faq/#are-there-any-limits (you get a single VM with 12 CPUs, so orchestration might be tricky in terms of isolation, cache race conditions, etc, but I haven't looked into it)
* https://www.bitrise.io/why/technologies/virtualized-m1-environment (we might need to reach out to see how we can be part of the trial)
* https://circleci.canny.io/cloud-feature-requests/p/support-new-m1-arm-based-macs (allegedly available by the end of January, and we have some infra for this already)

**Comment by jaimergp** [comment]
`conda/conda` experimenting with Cirrus M1 in this PR if you are curious on how it looks like: https://github.com/conda/conda/pull/12079

**Comment by adrianinsaval** [comment]
circleci has m1 machines available, is that a possibility?

**Comment by jakirkham** [comment]
GitHub posted a blogpost today about adding M1 workers

https://github.blog/2023-10-02-introducing-the-new-apple-silicon-powered-m1-macos-larger-runner-for-github-actions/

**Comment by adrianinsaval** [comment]
It's not clear to me if that's also free for open repos like the rest of the runners

**Comment by BastianZim** [comment]
No, they are larger runners that always cost something: https://docs.github.com/en/actions/using-github-hosted-runners/about-larger-runners/about-larger-runners#about-macos-larger-runners

**Comment by jaimergp** [comment]
The ticket where we might hear from Azure about adding M1 runners: https://github.com/actions/runner-images/issues/8971

**Comment by jaimergp** [comment]
Hey @conda-forge/core, I am curious about your opinion on the following idea: allowing selected feedstocks to run GHA `osx-arm64` runners, like we do with the GPU CI.

- Maintainers would open up a PR in `admin-requests` applying for access with a justification of need. A core member would review and see if their case is justified (mostly that they can't cross-compile, I'd assume).
- Once merged, they rerender their feedstock and they'd be allowed to use it.

To note:

- We don't want to this to DoS our infra jobs.
- I don't know (yet) how to implement this so people don't abuse it.
- How do we restrict the number of CI jobs a single feedstock can request? And globally?

**Comment by beckermr** [comment]
I think we'd only allow it if for cross-compiled builds are failing to work as opposed to first-come, first served.

**Comment by jaimergp** [comment]
Yes, I'd argue for a strong proof of "we have tried everything else and it doesn't work". Actually, the only feedstock I can think of is `conda-standalone` and even in that case it's ok to just build locally.

**Comment by xhochy** [comment]
We avoid native testing for all osx-arm64 feedstocks and only run the tests for the feedstock that are just existence tests (i.e. using `test -f`)

**Comment by jaimergp** [comment]
I might be reading this wrong, but [this page](https://learn.microsoft.com/en-us/azure/devops/pipelines/agents/hosted?view=azure-devops&tabs=yaml#software) hints that `macos-14` is available in preview in Azure now, which might mean arm64 runners?

> The macOS-14 Sonoma image is available in preview

**Comment by xhochy** [comment]
> I might be reading this wrong, but [this page](https://learn.microsoft.com/en-us/azure/devops/pipelines/agents/hosted?view=azure-devops&tabs=yaml#software) hints that `macos-14` is available in preview in Azure now, which might mean arm64 runners?
> 
> > The macOS-14 Sonoma image is available in preview

Immediately tried it out at https://github.com/conda-forge/mlx-feedstock/pull/35 and it looks like `amd64` runners.

**Comment by jaimergp** [comment]
Ah, pity. Yea, `uname -m` seems to return `x86_64`:

```
Installing a fresh version of Miniforge
+ MINIFORGE_URL=https://github.com/conda-forge/miniforge/releases/latest/download
++ uname -m
+ MINIFORGE_FILE=Mambaforge-MacOSX-x86_64.sh
```

**Comment by ilia-nikiforov-umn** [comment]
> daily(?) job testing published artefacts

Not directly related to this issue, but am I understanding correctly that, even though the `test` section of my `meta.yaml` is being skipped on `osx-arm64` in the feedstock, it is being run elsewhere after the package is built? If so, is there a place to see the output of these runs?

**Comment by beckermr** [comment]
It is NOT being run elsewhere.

**Comment by ilia-nikiforov-umn** [comment]
What does "testing published artefacts" entail, then?

**Comment by beckermr** [comment]
Download the artifact to an arm64 Mac. Then you can run conda build with tests only right on the artifact. I forget the exact command.

**Comment by h-vetinari** [comment]
> > daily(?) job testing published artefacts
> 
> If so, is there a place to see the output of these runs?

Note that when I wrote this, this _was_ actually my understanding. It's possible that I had misunderstood this (hence the "(?)"), but AFAIU there had been some effort to do more systematic testing as we ramped up osx-arm64 support. I had never seen it, but heard it spoken about (obviously I'm not going to find references to the exact comments years later). It's possible though that this never extended beyond some core developers doing manual checks.

**Comment by ilia-nikiforov-umn** [comment]
Got it, thank you. We will continue testing manually until hopefully native CI becomes available soon!

**Comment by beckermr** [comment]
We did a bunch of systematic testing at first to verify the initial cross-compiled builds worked. That was all done by hand, never automated. We don't currently have automated testing, even daily, setup. Given that we are close to native ci (hopefully) and that we'd likely see more bug reports from people using the packages live, I don't think we'll invest time in testing now. It was much more useful when there many fewer osx-arm64 users.

**Comment by hmaarrfk** [comment]
Are there any concerns in using github actions for native Arm-OSX Runners?

**Comment by beckermr** [comment]
We cannot do that. We use GitHub actions for backend service jobs that keep conda-forge running. If we have tons of ci jobs, things like the linter and admin webservices will stop working.

**Comment by beckermr** [comment]
Azure should have native osx arm soon.

**Comment by eunos-1128** [comment]
> Azure should have native osx arm soon.

Has a specific timeframe been announced?

I can't wait!

**Comment by beckermr** [comment]
This quarter. So hopefully before the end of the year.

**Comment by h-vetinari** [comment]
This is now working for a while; for completeness: you can switch from
```yaml
build_platform:
  osx_arm64: osx_64
```
to
```yaml
provider:
  osx_arm64: default
```
in `conda-forge.yml` and rerender.

Thanks everyone! I've opened a follow-up issue to have osx-arm64 builds by default (#2829), which should also allows us to avoid having to specify the `provider:` bit explicitly.
